pkg/nfs/controllerserver.go

```diff
@@ -73,3 +161,166 @@ func (cs *ControllerServer) ListSnapshots(ctx context.Context, req *csi.ListSnap
+	}
+
+	for _, c := range caps {
+		if err := cs.validateVolumeCapability(c); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (cs *ControllerServer) validateVolumeCapability(c *csi.VolumeCapability) error {
+	if c == nil {
+		return fmt.Errorf("volume capability must be provided")
+	}
+
+	// Validate access mode
+	accessMode := c.GetAccessMode()
+	if accessMode == nil {
+		return fmt.Errorf("volume capability access mode not set")
+	}
+	if !cs.Driver.cap[accessMode.Mode] {
+		return fmt.Errorf("driver does not support access mode: %v", accessMode.Mode.String())
+	}
+
+	// Validate access type
+	accessType := c.GetAccessType()
+	if accessType == nil {
+		return fmt.Errorf("volume capability access type not set")
+	}
+	mountType := c.GetMount()
+	if mountType == nil {
+		return fmt.Errorf("driver only supports mount access type volume capability")
+	}
+
+	if mountType.FsType != "" {
+		return fmt.Errorf("driver does not support fstype %v", mountType.FsType)
+	}
+	// TODO: check if we want to whitelist/blacklist certain mount options
+	return nil
+}
+
+// Mount nfs server at base-dir
+func (cs *ControllerServer) internalMount(ctx context.Context, vol *nfsVolume) error {
+	sharePath := filepath.Join("/" + vol.baseDir)
+	targetPath := cs.getInternalMountPath(vol)
+	stdVolCap := csi.VolumeCapability{
+		AccessType: &csi.VolumeCapability_Mount{
+			Mount: &csi.VolumeCapability_MountVolume{},
+		},
+	}
+
+	glog.V(4).Infof("internally mounting %v:%v at %v", vol.server, sharePath, targetPath)
+	_, err := cs.Driver.ns.NodePublishVolume(ctx, &csi.NodePublishVolumeRequest{
+		TargetPath: targetPath,
+		VolumeContext: map[string]string{
+			paramServer: vol.server,
+			attrShare:   sharePath,
+		},
+		VolumeCapability: &stdVolCap,
+		VolumeId:         vol.id,
+	})
+	return err
+}
+
+// Unmount nfs server at base-dir
+func (cs *ControllerServer) internalUnmount(ctx context.Context, vol *nfsVolume) error {
+	targetPath := cs.getInternalMountPath(vol)
+
+	// Unmount nfs server at base-dir
+	glog.V(4).Infof("internally unmounting %v", targetPath)
+	_, err := cs.Driver.ns.NodeUnpublishVolume(ctx, &csi.NodeUnpublishVolumeRequest{
+		TargetPath: cs.getInternalMountPath(vol),
+	})
+	return err
+}
+
+// Convert VolumeCreate parameters to an nfsVolume
+func (cs *ControllerServer) newNFSVolume(name string, params map[string]string) (*nfsVolume, error) {
+	var (
+		server  string
+		baseDir string
+	)
+
+	// Validate parameters (case-insensitive).
+	// TODO do more strict validation.
+	for k, v := range params {
+		switch strings.ToLower(k) {
+		case paramServer:
+			server = v
+		case paramBaseDir:
+			baseDir = v
+		default:
+			return nil, fmt.Errorf("invalid parameter %q", k)
+		}
+	}
+
+	// Validate required parameters
+	if server == "" {
+		return nil, fmt.Errorf("%v is a required parameter", paramServer)
+	}
+	if baseDir == "" {
+		return nil, fmt.Errorf("%v is a required parameter", paramBaseDir)
```

Comment:
@andyzhangx The sanity tests errors are coming here. If we want to keep this validation, we will have to update the existing sanity tests to use `server` and `baseDir` as request parameters. 
Is that ok?

Reply:
need to define a testing nfs-server in sanity test, otherwise even we ignore this error, `CreateVolume` would fail finally.

Reply:
And what would the values of these parameters be? Would adding these to the [defaults](func NewTestConfig() TestConfig {) break any existing tests?

Reply:
> need to set up a testing nfs-server in sanity test, otherwise even we ignore this error, CreateVolume would fail finally.

Can we run the [existing nfs server example](https://github.com/kubernetes-csi/csi-driver-nfs/tree/master/examples/kubernetes/nfs-provisioner) before initiating sanity tests? How would we feed in the `server` and `baseDir` parameters?

Reply:
I think yes, and I have modified the nfs server example: https://github.com/kubernetes-csi/csi-driver-nfs/pull/61, now could use `nfs-server.default.svc.cluster.local:/` in sanity test.

Need to add a flag in driver, if it's sanity test, use `nfs-server.default.svc.cluster.local:/`

Reply:
ah, in sanity test, there is no k8s cluster, it has only one machine, so you need to set up a containerized nfs-server, and use `127.0.0.1:/` for nfs-server.

Reply:
Seems like there's an issue with running docker containers within the sanity-test container:
```
docker: Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?.
```
Is docker in docker the right way to go about provisioning an nfs server for sanity tests?